app/components/billing/payment.js

```diff
@@ -53,6 +60,38 @@ export default Component.extend({
     }
   }).drop(),
 
+  // amount_off and price are in cents
+  discountedPrice: computed('coupon.{amountOff,percentOff}', 'selectedPlan.price', function () {
+    const price = Math.floor(this.selectedPlan.price / 100);
+    if (this.coupon && this.coupon.amountOff) {
+      const { amountOff } = this.coupon;
+      const discountedPrice = price - Math.floor(amountOff / 100);
+      return `$${discountedPrice}`;
+    } else if (this.coupon && this.coupon.percentageOff) {
+      const { percentageOff } = this.coupon;
+      const discountedPrice = price - (price * percentageOff) / 100;
+      return `$${discountedPrice.toFixed(2)}`;
+    } {
+      return `$${price}`;
+    }
+  }),
+
+  validateCoupon: task(function* () {
+    try {
+      const coupon = yield this.store.findRecord('coupon', this.couponId, {
+        reload: true,
+      });
+      this.set('coupon', coupon);
+    } catch (error) {
+      const containsCouponErrors = error && error.errors.length > 0;
+      if (containsCouponErrors) {
+        const couponError = error.errors[0];
+        const isInvalidCoupon = couponError.detail === `No such coupon: ${this.couponId}`;
```

Comment:
I would not rely on API error messages in UI. The message text can easily change in the future and the entire logic will break. Probably, any error would mean here that coupon is invalid (as long as we couldn't validate it), does it make sense? Or you could use HTTP error code to narrow the reason down.